File: python/datastructures/tree/heap/max_binary_heap.py
import math


class MaxBinaryHeap:

    def __init__(self):
        self.values = []

    def insert(self, value):
        self.values.push(value)

        idx = len(self.values) - 1
        parentIdx = math.floor((idx-1)/2)

        while self.values[idx] < self.values[parentIdx]:

            self.values[idx], self.values[parentIdx] = (
               self.values[parentIdx], self.values[idx]
               )

            idx = parentIdx
            parentIdx = math.floor(idx/2)

        return True

    def extractMax(self):

        max = self.values[0]

        end = self.values.pop()

        if len(self.values) > 0:

            self.values[0] = end

            self.sinkDown()

        return max

    def sinkDown(self):
        parent = 0
        # currentVal = self.values[0]

        while True:

            leftChildIdx = 2 * parent + 1

            rightChildIdx = 2 * parent + 2

            swap = None

            if leftChildIdx < len(self.values):

                if self.values[leftChildIdx] > self.values[parent]:
                    swap = leftChildIdx

            if rightChildIdx < len(self.values):

                if (
                    (
                        swap is None and self.values[rightChildIdx] >
                        self.values[parent]
                    )
                        or
                        (
                            swap is not None and self.values[rightChildIdx] >
                        self.leftChildIdx[parent])
                        ):
                    swap = rightChildIdx

            if swap is None:
                break

            self.values[parent], self.values[swap] = (
                    self.values[swap], self.values[parent])

            parent = swap

File: javascript/datastructures/binaryHeapsWithPriorityQueues/maxBinaryHeap.js
class MaxBinaryHeap{

    constructor(){
        this.values = [41,39,33,18,27,12]
    }

    insert(element){
        this.values.push(element)
        bubbleUp()        
    }

    bubbleUp(){
        let index = this.values.length - 1
        let parentIdx = Math.floor((index -1)/2)
        while(this.values[parentIdx] < this.values[index] && parentIdx > 0){
            [this.values[parentIdx], this.values[index]] = [this.values[index],this.values[parentIdx]]
            index = parentIdx 
            parentIdx = Math.floor((index - 1)/2)
        }
    }
    //removing element from a head
    extractMax() {
        let max = this.values[0]
        let end = this.values.pop()
        this.values[0] = end
        this.sinkDown()
        return max
    }

    sinkDown() {
        let idx = 0
        const length = this.values.length
        const element = this.values[0]
        let count = 0
        while (true) {
            if (count > 100)
                break
            count++
            let leftChildIdx = 2 * idx + 1
            let rightChildIdx = 2 * idx + 2
            let leftChild, rightChild
            let swap = null
            if (leftChildIdx < length) {
                leftChild = this.values[leftChildIdx]
                swap = leftChildIdx
                if (leftChild > element) {
                    swap = leftChildIdx
                }
            }
            if (rightChildIdx < length) {
                rightChild = this.values[rightChildIdx]
                if ((swap === null && rightChild > element) || (swap !== null && rightChild > leftChild)) {
                    swap = rightChildIdx
                }
            }
            if (swap === null)
                break
            this.values[idx] = this.values[swap]
            this.values[swap] = element
            idx = swap
        }
    }    
}
